pkg/ccl/changefeedccl/encoder_protobuf.go

```diff
@@ -0,0 +1,278 @@
+// Copyright 2025 The Cockroach Authors.
+//
+// Use of this software is governed by the CockroachDB Software License
+// included in the /LICENSE file.
+
+package changefeedccl
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/cdcevent"
+	"github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeedbase"
+	"github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeedpb"
+	"github.com/cockroachdb/cockroach/pkg/geo"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
+	"github.com/cockroachdb/errors"
+	"github.com/gogo/protobuf/types"
+)
+
+type protobufEncoder struct {
+	updatedField       bool
+	mvccTimestampField bool
+	beforeField        bool
+	keyInValue         bool
+	topicInValue       bool
+	envelopeType       changefeedbase.EnvelopeType
+	targets            changefeedbase.Targets
+}
+
+// protobufEncoderOptions wraps EncodingOptions for initializing a protobufEncoder.
+type protobufEncoderOptions struct {
+	changefeedbase.EncodingOptions
+}
+
+var _ Encoder = &protobufEncoder{}
+
+// newProtobufEncoder constructs a new protobufEncoder from the given options and targets.
+func newProtobufEncoder(
+	ctx context.Context, opts protobufEncoderOptions, targets changefeedbase.Targets,
+) Encoder {
+	return &protobufEncoder{
+		envelopeType:       opts.Envelope,
+		keyInValue:         opts.KeyInValue,
+		topicInValue:       opts.TopicInValue,
+		beforeField:        opts.Diff,
+		updatedField:       opts.UpdatedTimestamps,
+		mvccTimestampField: opts.MVCCTimestamps,
+		targets:            targets,
+	}
+}
+
+// EncodeKey serializes the primary key columns of a row as a changefeedpb.Key message.
+func (e *protobufEncoder) EncodeKey(ctx context.Context, row cdcevent.Row) ([]byte, error) {
+	keyMsg, err := buildKeyMessage(row)
+	if err != nil {
+		return nil, err
+	}
+	return protoutil.Marshal(keyMsg)
+}
+
+// EncodeValue serializes a row event using the configured envelope type.
+func (e *protobufEncoder) EncodeValue(
+	ctx context.Context, evCtx eventContext, updatedRow, prevRow cdcevent.Row,
+) ([]byte, error) {
+	switch e.envelopeType {
+	case changefeedbase.OptEnvelopeBare:
+		return e.buildBare(evCtx, updatedRow, prevRow)
+	default:
+		return nil, errors.AssertionFailedf("envelope format not supported: %s", e.envelopeType)
+	}
+}
+
+// EncodeResolvedTimestamp encodes a resolved timestamp message for the specified topic.
+func (e *protobufEncoder) EncodeResolvedTimestamp(
+	context.Context, string, hlc.Timestamp,
+) ([]byte, error) {
+	return nil, errors.Newf("protobuf encoder does not support resolved timestamps yet — see https://github.com/cockroachdb/cockroach/issues/148934")
```

Comment:
technically we have an Unimplemented error helper for this... but it's fine since this work is next. but fyi